src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,920 @@
+ * Example: SourceShard is range [0, 100] and SplitPoints are (15, 30) will give us:
+ *  [(0, 15) (16, 30) (31, 100)]
+ */
+static void
+CreateSplitIntervalsForShard(ShardInterval *sourceShard,
+							 List *splitPointsForShard,
+							 List **shardSplitChildrenIntervalList)
+{
+	/* For 'N' split points, we will have N+1 shard intervals created. */
+	int shardIntervalCount = list_length(splitPointsForShard) + 1;
+	ListCell *splitPointCell = list_head(splitPointsForShard);
+	int32 splitParentMaxValue = DatumGetInt32(sourceShard->maxValue);
+
+	int32 currentSplitChildMinValue = DatumGetInt32(sourceShard->minValue);
+	for (int index = 0; index < shardIntervalCount; index++)
+	{
+		ShardInterval *splitChildShardInterval = CopyShardInterval(sourceShard);
+		splitChildShardInterval->shardIndex = -1;
+		splitChildShardInterval->shardId = GetNextShardId();
+
+		splitChildShardInterval->minValueExists = true;
+		splitChildShardInterval->minValue = currentSplitChildMinValue;
+		splitChildShardInterval->maxValueExists = true;
+
+		/* Length of splitPointsForShard is one less than 'shardIntervalCount' and we need to account */
+		/* for 'splitPointCell' being NULL for last iteration. */
+		if (splitPointCell)
+		{
+			splitChildShardInterval->maxValue = DatumGetInt32((Datum) lfirst(
+																  splitPointCell));
+			splitPointCell = lnext(splitPointsForShard, splitPointCell);
+		}
+		else
+		{
+			splitChildShardInterval->maxValue = splitParentMaxValue;
+		}
+
+		currentSplitChildMinValue = splitChildShardInterval->maxValue + 1;
+		*shardSplitChildrenIntervalList = lappend(*shardSplitChildrenIntervalList,
+												  splitChildShardInterval);
+	}
+}
+
+
+/*
+ * Insert new shard and placement metadata.
+ * Sync the Metadata with all nodes if enabled.
+ */
+static void
+InsertSplitChildrenShardMetadata(List *shardGroupSplitIntervalListList,
+								 List *workersForPlacementList)
+{
+	/* Iterate on shard intervals for shard group */
+	List *shardIntervalList = NULL;
+	List *syncedShardList = NULL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		/* Iterate on split children shards along with the respective placement workers */
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			InsertShardRow(
+				shardInterval->relationId,
+				shardInterval->shardId,
+				shardInterval->storageType,
+				IntegerToText(DatumGetInt32(shardInterval->minValue)),
+				IntegerToText(DatumGetInt32(shardInterval->maxValue)));
+
+			InsertShardPlacementRow(
+				shardInterval->shardId,
+				INVALID_PLACEMENT_ID, /* triggers generation of new id */
+				SHARD_STATE_ACTIVE,
+				0, /* shard length (zero for HashDistributed Table) */
+				workerPlacementNode->groupId);
+		}
+
+		if (ShouldSyncTableMetadata(shardInterval->relationId))
```

Comment:
Found this bug with testing: the if check should be inside the forboth_ptr loop.